### PR TITLE
feat: implement API gateway topology at the dataplane and environment level

### PR DESCRIPTION
**Pull request by NomadXD**
## Purpose
This PR implements the API gateway topology at the dataplane and environment level. 

## Related Issues
https://github.com/openchoreo/openchoreo/issues/1797

## Checklist
- [ ] Tests added or updated (unit, integration, etc.)
- [ ] Samples updated (if applicable)

API Gateway Topology Implementation at Dataplane and Environment Level

1) Mini breakdown — changed files by top-level folder (from supplied diffs)
- api/: 3 files
- config/: 4 files (config/crd/bases/)
- install/: 6 files (install/helm/... CRD manifests)
- internal/: ~18 files (controller, pipeline/context, occ CLI, tests)
- samples/: 2 files
- other: small edits in repo manifests/tests (e.g., go.mod, tests)

2) API / CRD surface changes (what changed, compatibility risk)
- Additive gateway identifier fields — Low risk
  - New GatewaySpec fields added (DataPlane / ClusterDataPlane / Environment CRDs): publicGatewayName, publicGatewayNamespace, organizationGatewayName, organizationGatewayNamespace (defaults: "gateway-default", "openchoreo-data-plane"). Additive with defaults → compatibility risk: Low.
- Environment gateway schema redesign — High risk
  - Removed: gateway.dnsPrefix and gateway.security (remoteJwks.uri); removed public types GatewayConfig, SecurityConfig, RemoteJWKS; EnvironmentSpec.Gateway type changed from GatewayConfig → GatewaySpec.
  - Added/changed: explicit fields under spec.gateway: publicVirtualHost (now required), organizationVirtualHost (optional), public/organization HTTP/HTTPS ports, plus the 4 gateway identifier fields.
  - Compatibility risk: High — removal of types/fields and a new required field will break existing manifests/typed clients without migration.
- ReleaseBinding status addition — Low risk
  - New status type EndpointURLStatus { name, invokeURL } and ReleaseBindingStatus.Endpoints []EndpointURLStatus added. Additive status field → compatibility risk: Low.

3) Controller / behavioral changes
- ReleaseBinding reconciler: endpoint URL resolution and status population
  - Adds unstructured HTTPRoute parsing, backendRef→workload endpoint matching, and URL composition (host + optional path + port rules). Populates ReleaseBinding.Status.Endpoints and clears it on undeploy.
  - Port precedence/fallback: environment override → dataplane → standard ports; supports non-standard ports (included in URL).
  - New helpers: extractBackendRefPort, extractFirstHostname, extractFirstPathValue, extractFirstParentRef, resolveGatewayPort, isStandardPort. Significant reconciling logic added (~+271 LOC).
  - Logging added for resolution flow.
- Pipeline/context & templating
  - New EnvironmentData type (PublicVirtualHost, OrganizationVirtualHost) and propagation into ComponentContext, TraitContext, embedded trait contexts, and Pipeline.Render; DataPlaneData extended with OrganizationVirtualHost.
- CLI / API surfaces
  - occ CLI: removed DNSPrefix column; environment creation no longer sets Gateway.DNSPrefix.
  - internal/openchoreo-api: no longer populates DNSPrefix in EnvironmentResponse.

4) Tests added/updated (and gaps)
- Tests updated
  - Controller/unit tests modified to remove gateway.dnsPrefix usage (several controller tests).
  - Pipeline/context builder and embedded trait tests updated to include Environment in inputs and expected outputs.
  - Samples updated to remove gateway.dnsPrefix from three Environment manifests.
- Tests added
  - New comprehensive unit tests for endpoint resolution: internal/controller/releasebinding/endpoint_resolve_test.go (exercises resolveEndpointURLStatuses and helpers across many scenarios).
- Remaining critical test gaps
  - Ensure integration/e2e tests that verify ReleaseBinding.Status.Endpoints population end-to-end (controller + Gateway API resources + rendered HTTPRoutes) exist — not shown in diffs.
  - Verify upgrade/CRD migration e2e tests to validate manifest conversion from dnsPrefix/security to new schema.

5) Risk hotspots (short reasons)
- Backward compatibility & upgrade/migration — HIGH
  - Removal of gateway.dnsPrefix, security block, and public types; new required publicVirtualHost. Existing manifests and typed clients will break; migration guidance required.
- Install/upgrade path & CRD rollout — MEDIUM–HIGH
  - CRD schema changes (removals + new required field) may cause validation failures during Helm/operator upgrades; ensure CRDs applied before controller rollout and provide a migration plan.
- Reconciliation complexity — MEDIUM
  - Unstructured HTTPRoute parsing increases surface area and susceptibility to Gateway API schema variants and edge cases (multiple hostnames/parents/paths).
- RBAC / API access — MEDIUM
  - New controller logic reads Gateway API kinds (HTTPRoute/Gateway); RBAC manifests were not shown — add list/get/watch permissions for Gateway API resources if missing.
- AuthN/AuthZ & secrets handling — MEDIUM
  - Removal of remoteJwks/security from Environment CRD — confirm where JWKS/security now resides and ensure equivalent auth config is handled; risk to authentication flows if overlooked.
- Logging & observability — LOW
  - Added logs aid debugging but review verbosity/levels.

6) Summary statistics and notable diffs (from provided diffs)
- Largest code changes:
  - internal/controller/releasebinding/controller.go: ~+271 LOC (endpoint resolution logic and helpers).
  - internal/pipeline/component/context/*: multiple files updated to add EnvironmentData and propagation (combined +~150 LOC across types, builder, tests).
  - CRD schema changes:
    - config/crd/bases/openchoreo.dev_environments.yaml: +49 / -11 (environment gateway redesign — largest CRD change).
    - config/crd/bases/openchoreo.dev_dataplanes.yaml & openchoreo.dev_clusterdataplanes.yaml: each ~+20 / -1 (gateway identifier fields).
    - config/crd/bases/openchoreo.dev_releasebindings.yaml: +22 / 0 (status.endpoints added).
- Samples: removed gateway.dnsPrefix from samples/getting-started/all.yaml and samples/getting-started/environments.yaml.

Actionable reviewer checklist
- Validate CRD upgrade/migration plan and ordering; provide migration docs and compatibility guidance for dnsPrefix/security removal and GatewayConfig type removal.
- Ensure operator Helm/CRD manifests apply new CRDs before controller rollout; consider compatibility shim or staged rollout if needed.
- Add/verify RBAC rules for Gateway API kinds (HTTPRoute, Gateway) in controller ClusterRole/Role.
- Confirm where JWKS/security configuration is intended to live now that Environment CRD removed remoteJwks/security block; ensure authn/authz remains covered.
- Review added reconciliation logic for edge-case handling and add integration/e2e tests to assert ReleaseBinding.Status.Endpoints population end-to-end.
- Consider log level adjustments if resolution logging is too verbose in normal operation.

### CODE REVIEW

File: api/v1alpha1/environment_types.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ type EnvironmentSpec struct {
 	// Immutable once set.
 	DataPlaneRef *DataPlaneRef `json:"dataPlaneRef,omitempty"`
 	IsProduction bool          `json:"isProduction,omitempty"`
-	Gateway      GatewayConfig `json:"gateway,omitempty"`
+	Gateway      GatewaySpec   `json:"gateway,omitempty"`
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply api/v1alpha1/environment_types.go:21]: `GatewayConfig` is optional and not currently used and it was used in the initial implementation of OpenChoreo.

File: install/helm/openchoreo-control-plane/crds/openchoreo.dev_releasebindings.yaml [hunk 2]
```diff
@@ -290,6 +290,28 @@ spec:
...
+                      type: string
+                    name:
+                      description: Name is the endpoint name as defined in the Workload
+                        spec.
+                      type: string
+                  required:
+                  - invokeURL
+                  - name
+                  type: object
+                type: array
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply install/helm/openchoreo-control-plane/crds/openchoreo.dev_releasebindings.yaml:314]: auto generated CRD files. hence resolving

File: internal/controller/releasebinding/controller.go [hunk 3]
```diff
@@ -916,6 +937,281 @@ func (r *Reconciler) generateResourceID(resource map[string]any, index int) stri
...
+	if len(endpoints) == 0 {
+		logger.Info("No workload endpoints defined, skipping endpoint URL resolution")
+		return nil
+	}
+
+	// Build port → endpoint name lookup (port is unique per workload)
+	portToName := make(map[int64]string, len(endpoints))
+	for name, ep := range endpoints {
+		portToName[int64(ep.Port)] = name
+		logger.Info("Registered workload endpoint", "name", name, "port", ep.Port, "type", ep.Type)
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/controller/releasebinding/controller.go:963]: Uniqueness should be enforced at the Workload CR level. This will be handled with the current Endpoint API revisit task.

File: internal/pipeline/component/context/component.go [hunk 4]
```diff
@@ -182,6 +182,61 @@ func extractDataPlaneData(dp *v1alpha1.DataPlane) DataPlaneData {
...
+	}
+	return name
+}
+
+func gatewayNamespaceOrDefault(namespace string) string {
+	if namespace == "" {
+		return defaultGatewayNamespace
+	}
+	return namespace
+}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/pipeline/component/context/component.go]: fixed with [feat: add environment to component context](https://github.com/openchoreo/openchoreo/pull/1902/commits/e9289aab5ce69e868b1a8cb254ebb79189c23e27)

File: internal/pipeline/component/context/types.go [hunk 5]
```diff
@@ -195,6 +195,21 @@ type ObservabilityPlaneRefData struct {
...
+	OrganizationVirtualHost      string `json:"organizationVirtualHost,omitempty"`
+	PublicHTTPPort               int32  `json:"publicHTTPPort,omitempty"`
+	PublicHTTPSPort              int32  `json:"publicHTTPSPort,omitempty"`
+	OrganizationHTTPPort         int32  `json:"organizationHTTPPort,omitempty"`
+	OrganizationHTTPSPort        int32  `json:"organizationHTTPSPort,omitempty"`
+	PublicGatewayName            string `json:"publicGatewayName,omitempty"`
+	PublicGatewayNamespace       string `json:"publicGatewayNamespace,omitempty"`
+	OrganizationGatewayName      string `json:"organizationGatewayName,omitempty"`
+	OrganizationGatewayNamespace string `json:"organizationGatewayNamespace,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/pipeline/component/context/types.go:217]: fixed with [feat: add environment to component context](https://github.com/openchoreo/openchoreo/pull/1902/commits/e9289aab5ce69e868b1a8cb254ebb79189c23e27)

File: internal/pipeline/component/context/component.go [hunk 6]
```diff
@@ -182,6 +186,61 @@ func extractDataPlaneData(dp *v1alpha1.DataPlane) DataPlaneData {
...
+)
+
+// extractEnvironmentData extracts EnvironmentData from Environment and DataPlane resources.
+// If the Environment has gateway configuration, it uses those values.
+// Otherwise, it falls back to the DataPlane gateway configuration.
+// Gateway name and namespace default to "gateway-default" and "openchoreo-data-plane" if not set.
+func extractEnvironmentData(env *v1alpha1.Environment, dp *v1alpha1.DataPlane) EnvironmentData {
+	// If environment has gateway configuration, use it
+	if env.Spec.Gateway.PublicVirtualHost != "" {
+		return EnvironmentData{
```
Comment (ChathurangaKCD) [inline internal/pipeline/component/context/component.go:196]: EnvironmentData or GatewayInfo/GtewayConfig?

File: internal/pipeline/component/context/component.go [hunk 7]
```diff
@@ -182,6 +186,61 @@ func extractDataPlaneData(dp *v1alpha1.DataPlane) DataPlaneData {
...
+		PublicHTTPPort:               dp.Spec.Gateway.PublicHTTPPort,
+		PublicHTTPSPort:              dp.Spec.Gateway.PublicHTTPSPort,
+		OrganizationHTTPPort:         dp.Spec.Gateway.OrganizationHTTPPort,
+		OrganizationHTTPSPort:        dp.Spec.Gateway.OrganizationHTTPSPort,
+		PublicGatewayName:            gatewayNameOrDefault(dp.Spec.Gateway.PublicGatewayName),
+		PublicGatewayNamespace:       gatewayNamespaceOrDefault(dp.Spec.Gateway.PublicGatewayNamespace),
+		OrganizationGatewayName:      gatewayNameOrDefault(dp.Spec.Gateway.OrganizationGatewayName),
+		OrganizationGatewayNamespace: gatewayNamespaceOrDefault(dp.Spec.Gateway.OrganizationGatewayNamespace),
+	}
+}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/pipeline/component/context/component.go]: With the current implementation Public gateway is mandatory and Organization gateway is optional. So just checking the Public gateway is sufficient with the current implementation.

File: internal/pipeline/component/context/types.go [hunk 8]
```diff
@@ -184,9 +197,10 @@ type ComponentContext struct {
 
 // DataPlaneData provides data plane configuration in templates.
 type DataPlaneData struct {
-	SecretStore           string                     `json:"secretStore,omitempty"`
-	PublicVirtualHost     string                     `json:"publicVirtualHost,omitempty"`
-	ObservabilityPlaneRef *ObservabilityPlaneRefData `json:"observabilityPlaneRef,omitempty"`
+	SecretStore             string                     `json:"secretStore,omitempty"`
+	PublicVirtualHost       string                     `json:"publicVirtualHost,omitempty"`
+	OrganizationVirtualHost string                     `json:"organizationVirtualHost,omitempty"`
```
Comment (ChathurangaKCD) [inline internal/pipeline/component/context/types.go:202]: Now that both fields are in EnvironmentData, do we need to keep it here?

File: internal/pipeline/component/context/component.go (hunk 6 above)
Comment (NomadXD) [reply internal/pipeline/component/context/component.go:196]: @ChathurangaKCD I removed all the fields except public and organization vhost with [feat: remove port and gateway name from EnvironmentData](https://github.com/openchoreo/openchoreo/pull/1902/commits/f0610a89c1f1744ed9632a87997e74347501523e). We can keep it as EnvironmentData similar to DataplaneData and add other non gateway related environment properties as well as required in future.

File: internal/pipeline/component/context/types.go (hunk 8 above)
Comment (NomadXD) [reply internal/pipeline/component/context/types.go:202]: @ChathurangaKCD With the gateway topology feature, a gateway can be specified at the dataplane level or the environment level. The default topology mode is a gateway at the dataplane level. In the ComponentType, users can use both `dataplane. PublicVirtualHost` and `environment. PublicVirtualHost` depending on the configured gateway topology. `environment. PublicVirtualHost` will resolve to `dataplane. PublicVirtualHost` if no environment level gateway is configured. But there can be cases where the users will need to specify `dataplane.PublicVirtualHost` explicitly in a scenario where we have configured an environment level gateway. Hence we need the vhosts at both level.

File: internal/controller/releasebinding/controller.go [hunk 9]
```diff
@@ -916,6 +937,256 @@ func (r *Reconciler) generateResourceID(resource map[string]any, index int) stri
...
+			"parentRefName", parentRefName,
+			"parentRefNamespace", parentRefNamespace,
+			"path", path,
+		)
+
+		result = append(result, openchoreov1alpha1.EndpointURLStatus{
+			Name:      endpointName,
+			InvokeURL: invokeURL,
+		})
+	}
```
[the remark this answers is not in the document]
Comment (NomadXD) [reply internal/controller/releasebinding/controller.go:1035]: Unit tests are added with [feat: add unit tests for endpoint url resolution](https://github.com/openchoreo/openchoreo/pull/1902/commits/41ee5d30d06656e861e6459112c3c4f5c7463c8d)
